core/src/main/java/com/linecorp/armeria/internal/common/stream/AbstractFixedStreamMessage.java

```diff
@@ -34,60 +27,43 @@
  * {@link EmptyFixedStreamMessage}, {@link OneElementFixedStreamMessage}, or
  * {@link TwoElementFixedStreamMessage} when publishing less than three objects.
  */
-@UnstableApi
-public class RegularFixedStreamMessage<T> extends FixedStreamMessage<T> {
-
-    private final T[] objs;
+abstract class AbstractFixedStreamMessage<T> extends FixedStreamMessage<T> {
 
     private int fulfilled;
     private boolean inOnNext;
     private boolean cancelled;
 
     private volatile int demand;
 
-    /**
-     * Creates a new instance with the specified elements.
-     */
-    protected RegularFixedStreamMessage(T[] objs) {
-        requireNonNull(objs, "objs");
-        for (int i = 0; i < objs.length; i++) {
-            if (objs[i] == null) {
-                throw new NullPointerException("objs[" + i + "] is null");
-            }
-        }
-
-        this.objs = objs.clone();
-    }
-
     @Override
-    public long demand() {
+    public final long demand() {
         return demand;
     }
 
+    abstract T get(int index);
+
+    abstract int size();
+
     @Override
     final void cleanupObjects(@Nullable Throwable cause) {
-        while (fulfilled < objs.length) {
-            final T obj = objs[fulfilled];
-            objs[fulfilled++] = null;
```

Comment:
IIUC, it doesn't violate Reactive Streams 3.13 because it does mean elements of a stream.
We remove a `Subscriber` reference from a `Publisher` to abide by 3.13.
https://github.com/line/armeria/blob/8fd9663aa3ef90c897a233225a28cce1058f2b40/core/src/main/java/com/linecorp/armeria/common/stream/AbstractStreamMessage.java#L176-L184
https://github.com/line/armeria/blob/a55279039d45740b0cfe44e81945ae62fc1cc29b/core/src/main/java/com/linecorp/armeria/internal/common/stream/FixedStreamMessage.java#L289

Reply:
I see, I supposer the messages will bc cleaned up when the subscriber is gc-ed (and the messages are already aggregated in memory) so it doesn't matter. Sounds good to me 👍